File: src/handlers/auth_handler.rs
use axum::{
    Json,
    extract::Extension,
    response::{IntoResponse, Response},
};
use tracing::instrument;

use crate::models::user::{CreateUser, LoginRequest};
use crate::services::Services;

#[instrument(skip(services, payload))]
pub async fn register(
    Extension(services): Extension<Services>,
    Json(payload): Json<CreateUser>,
) -> Response {
    let user = match services
        .auth
        .register(&payload.username, &payload.email, &payload.password)
        .await
    {
        Ok(u) => u,
        Err(e) => return e.into_response(),
    };

    let token = match services.auth.login(&payload.username, &payload.password).await {
        Ok((_, t)) => t,
        Err(e) => return e.into_response(),
    };

    services.auth.build_register_response(user, &token)
}

#[instrument(skip(services, payload))]
pub async fn login(
    Extension(services): Extension<Services>,
    Json(payload): Json<LoginRequest>,
) -> Response {
    let (user, token) = match services.auth.login(&payload.username, &payload.password).await {
        Ok((u, t)) => (u, t),
        Err(e) => return e.into_response(),
    };

    services.auth.build_auth_cookies_response(user, &token)
}

pub async fn logout(services: Extension<Services>) -> Response {
    services.auth.build_logout_response()
}

#[instrument(skip(services))]
pub async fn me(
    services: Extension<Services>,
    auth: crate::middleware::auth::AuthenticatedUser,
) -> Response {
    match services.auth.get_user(&auth.user_id).await {
        Ok(Some(user)) => {
            use axum::{Json, http::StatusCode};
            (StatusCode::OK, Json(user)).into_response()
        }
        Ok(None) => {
            use axum::{Json, http::StatusCode};
            let body = serde_json::json!({"error": "User not found"});
            (StatusCode::NOT_FOUND, Json(body)).into_response()
        }
        Err(e) => e.into_response(),
    }
}

File: apps/backend-go/internal/adapters/http/auth_handler.go
package httpadapter

import (
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"

	"github.com/velopulent/cms/apps/backend-go/internal/application"
	"github.com/velopulent/cms/apps/backend-go/internal/domain"
	"github.com/velopulent/cms/apps/backend-go/internal/middleware"
)

type AuthHandler struct {
	auth         *application.AuthService
	cookieSecure bool
}

func NewAuthHandler(auth *application.AuthService, cookieSecure bool) AuthHandler {
	return AuthHandler{auth: auth, cookieSecure: cookieSecure}
}

type createUserRequest struct {
	Username string `json:"username"`
	Email    string `json:"email"`
	Password string `json:"password"`
}

type loginRequest struct {
	Username string `json:"username"`
	Password string `json:"password"`
}

func (h AuthHandler) Register(c echo.Context) error {
	var req createUserRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, errorBody{Error: "Invalid input"})
	}

	user, token, err := h.auth.Register(c.Request().Context(), req.Username, req.Email, req.Password)
	if err != nil {
		return writeAppError(c, err)
	}

	h.setAuthCookies(c, token)
	return c.JSON(http.StatusCreated, domain.AuthResponse{User: user})
}

func (h AuthHandler) Login(c echo.Context) error {
	var req loginRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, errorBody{Error: "Invalid input"})
	}

	user, token, err := h.auth.Login(c.Request().Context(), req.Username, req.Password)
	if err != nil {
		return writeAppError(c, err)
	}

	h.setAuthCookies(c, token)
	return c.JSON(http.StatusOK, domain.AuthResponse{User: user})
}

func (h AuthHandler) Logout(c echo.Context) error {
	h.clearAuthCookies(c)
	return c.JSON(http.StatusOK, map[string]string{"message": "Logged out"})
}

func (h AuthHandler) Me(c echo.Context) error {
	userID, _ := c.Get(middleware.UserIDKey).(string)
	user, err := h.auth.GetUser(c.Request().Context(), userID)
	if err != nil {
		return writeAppError(c, err)
	}
	if user == nil {
		return c.JSON(http.StatusNotFound, errorBody{Error: "User not found"})
	}
	return c.JSON(http.StatusOK, user)
}

func (h AuthHandler) setAuthCookies(c echo.Context, token string) {
	maxAge := int((24 * time.Hour).Seconds())
	c.SetCookie(&http.Cookie{
		Name:     "token",
		Value:    token,
		Path:     "/",
		MaxAge:   maxAge,
		HttpOnly: true,
		Secure:   h.cookieSecure,
		SameSite: http.SameSiteStrictMode,
	})
	c.SetCookie(&http.Cookie{
		Name:     "csrf",
		Value:    uuid.Must(uuid.NewV7()).String(),
		Path:     "/",
		MaxAge:   maxAge,
		HttpOnly: false,
		Secure:   h.cookieSecure,
		SameSite: http.SameSiteStrictMode,
	})
}

func (h AuthHandler) clearAuthCookies(c echo.Context) {
	c.SetCookie(&http.Cookie{
		Name:     "token",
		Value:    "",
		Path:     "/",
		MaxAge:   -1,
		HttpOnly: true,
		SameSite: http.SameSiteDefaultMode,
	})
	c.SetCookie(&http.Cookie{
		Name:     "csrf",
		Value:    "",
		Path:     "/",
		MaxAge:   -1,
		HttpOnly: false,
		SameSite: http.SameSiteDefaultMode,
	})
}
